Modify rule S1994: promote C# to SonarWay (#3694)
SonarSource/rspec: rules/S1994/csharp/rule.adoc

== Why is this an issue?

The `for` loop is designed to iterate over a range using a counter variable, with the counter being updated in the loop's increment section. Misusing this structure can lead to issues such as infinite loops if the counter is not updated correctly.

> Using a for loop for purposes other than its intended use can lead to confusion and potential bugs. If the `for` loop structure does not fit your needs, consider using an alternative https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements[iteration statement] should be considered in those cases.

Reviewer: "consider using an alternative iteration statement should be considered in those cases."
I think you copy pasted my ugly sentence full of mistakes here, thanks for the trust 😄. My bad sorry, can you rewrite it in proper English?

> If the counter is not updated, the loop will be infinite with a constant counter variable. If this is intentional, use a `while` or `do while` loop instead of a `for` loop.

Reviewer: Should this sentence stay? It might be redundant with the one above, not a strong opinion

Author: I like the suggestion. I would keep, I don't find it redundant.

Author: After, rereading I move the suggestion to the first paragraph.

== How to fix it

Move the counter variable update to the loop's increment section.

> 

Reviewer: This empty line might also be unnecessary with such short sentences

Author: I want to keep the two solutions on separate lines, I don't keep the empty line, it will be on the same line.

If this is impossible, consider using another iteration statement instead.

=== Code examples

==== Noncompliant code example

[source,csharp,diff-id=1,diff-type=noncompliant]
----
int sum = 0;

> for (int i = 0; i < 10; sum++) // Noncompliant, i not updated in increment section

Reviewer: ```suggestion
for (int i = 0; i < 10; sum++) // Noncompliant: `i` is not updated in the increment section
```

{
  // ...
  i++;
}
----

[source,csharp,diff-id=2,diff-type=noncompliant]
----

> for (int i = 0;; i++) // Noncompliant, the loop condition is empty althoug incrementing i

Reviewer: ```suggestion
for (int i = 0;; i++) // Noncompliant, the loop condition is empty although incrementing `i`
```

{
  // ...
}
----

==== Compliant solution

[source,csharp,diff-id=1,diff-type=compliant]
----
int sum = 0;

> for (int i = 0; i < 10; i++) // Compliant

Reviewer: ```suggestion
for (int i = 0; i < 10; i++)
```

{
  // ...
  sum++;
}
----

[source,csharp,diff-id=2,diff-type=compliant]
----
int i = 0;

> while (true) // Compliant

Reviewer: ```suggestion
while (true)
```

== Why is this an issue?

> The counter of a `for` loop should be updated in the loop's increment section.
> The purpose of a `for` loop is to iterate over a range using a counter variable.
> It should not be used for other purposes, and alternative https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements[iteration statement] should be considered in those cases.

Reviewer: I don't find this why section particularly fluent, specifically the first two "bullet point" sentences.
What about:
```suggestion
The `for` loop is designed to iterate over a range using a counter variable, with the counter being updated in the loop's increment section. Misusing this structure can lead to issues such as infinite loops if the counter is not updated correctly. 

Using a for loop for purposes other than its intended use can lead to confusion and potential bugs. If the `for` loop structure does not fit your needs, consider using an alternative https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements[iteration statement].
```
Partial courtesy of chat geppetto.

If the counter is not updated, the loop will be infinite with a constant counter variable. If this is intentional, use a `while` or `do while` loop instead of a `for` loop.

== How to fix it

Move the counter variable update to the loop's increment section.

If this is impossible, consider using another iteration statement instead.

> === Code examples

Reviewer: IMO we can improve a bit the consistency of this section: we have 2 snippets within a `==== Noncompliant code example` with their counterparts in `==== Compliant solution` and then another `==== Noncompliant code example` with 1 snippet and it's counterpart. What is the difference between the first 2 and the third example? 
Now:
```
- Code examples
----- Noncompliant section
---------- 1 Noncompliant example
---------- 2 Noncompliant example
----- Compliant section
---------- 1 Compliant example
---------- 2 Compliant example
----- Noncompliant section
---------- 3 Noncompliant example
----- Compliant section
---------- 3 Compliant example
```
Either:
```
- Code examples
----- Noncompliant section
---------- 1 Noncompliant example
---------- 2 Noncompliant example
---------- 3 Noncompliant example
----- Compliant section
---------- 1 Compliant example
---------- 2 Compliant example
---------- 3 Compliant example
```
or:
```
- Code examples
----- Noncompliant section
---------- 1 Noncompliant example
----- Noncompliant section
---------- 2 Noncompliant example
----- Noncompliant section
---------- 3 Noncompliant example
----- Compliant section
---------- 1 Compliant example
----- Compliant section
---------- 2 Compliant example
----- Compliant section
---------- 3 Compliant example
```
I would go with the first one.

Reviewer: And I would suggest also cutting one example, 2 seem like the sweet spot

Author: I think the second suggestion is not allowed by LaYC rules.
My idea was to have a clear separation between the example of the two sentences in the `How to fix it` section.

I will apply your suggestion of 2 examples.

==== Noncompliant code example

[source,csharp,diff-id=1,diff-type=noncompliant]
----
int sum = 0;
for (int i = 0; i < 10; sum++) // Noncompliant, i not updated in increment section
{
  // ...
  i++;
}
----

[source,csharp,diff-id=2,diff-type=noncompliant]
----

> for (int i = 0;; i++) // Noncompliant, the loop condition is empty althoug incrementing i

Reviewer: I think the purpose of a dev writing this code would be to have an infinite loop with an infinite incrementer. What about having a `while loop` as the compliant example?
